### ISSUE TITLE
revalidation blocks page load on dynamic routes

### ISSUE THREAD

**Issue by mattddean** [vercel/next.js#53324]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.15.1
      npm: 8.13.2
      Yarn: 1.22.19
      pnpm: 8.6.7
    Relevant Packages:
      next: 13.4.13-canary.6
      eslint-config-next: 13.4.13-canary.6
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/mattddean/nextjs-boilerplate-caching

### To Reproduce

1. Load the homepage of the [deployed app](https://nextjs-boilerplate-caching-cnp48vetr-hello-bello.vercel.app/). Observe that it takes 20 seconds to load.
2. Refresh the page. Observe that it loads instantly.
3. Keep refreshing the page through 10 seconds, at which point page load will be slow again.
4. Refresh the page again. Observe that it loads instantly.

### Describe the Bug

unstable_cache, when used with its `revalidate` option, blocks page load when revalidating instead of performing the revalidation in the background.

### Expected Behavior

Stale data should be retrieved from unstable_cache until revalidation has occurred in the background.

### Which browser are you using? (if relevant)

Chrome 115.0.5790.114

### How are you deploying your application? (if relevant)

Vercel

**Comment by mattddean** [comment]
Seems this isn't limited to unstable_cache. It's also the behavior of fetch for dynamic routes. This behavior seems to conflict with [these docs](https://nextjs.org/docs/app/building-your-application/caching#time-based-revalidation). Stale data is not returned after the revalidation interval. Instead, revalidation happens while the user waits for the page to load.

https://nextjs-boilerplate-caching-n733y81yh-hello-bello.vercel.app/fetch

**Comment by wuifdesign** [comment]
Any update/workaround for this? i'm having the same problem using fetch.

also made a repo for reproduction https://github.com/wuifdesign/next-js-test

**Comment by david-keefe** [comment]
Isn't this the expected behavior? If the revalidation limit has been exceeded, it fetches the data again. Otherwise you'd be serving a stale page.

**Comment by mattddean** [comment]
@david-keefe 
> Isn't this the expected behavior? If the revalidation limit has been exceeded, it fetches the data again. Otherwise you'd be serving a stale page.

Yes, but the unexpected thing to me was when the revalidation takes place. Contrary to what these docs seem to say, for dynamic routes, instead of taking place in the background, it takes place on the user's time while they wait for their data to load.

https://nextjs.org/docs/app/building-your-application/caching#time-based-revalidation

>After the timeframe, the next request will still return the cached (now stale) data.
>Next.js will trigger a revalidation of the data in the background.
>Once the data is fetched successfully, Next.js will update the Data Cache with the fresh data.
>If the background revalidation fails, the previous data will be kept unaltered.

**Comment by wuifdesign** [comment]
moreover it takes the user's time to revalidate the cache but still returns the old data. the new data is only retuned in the next request done.

**Comment by mattddean** [comment]
Looks like this is fixed by #55978, but I haven't tested it yet

**Comment by mattddean** [comment]
> Looks like this is fixed by #55978, but I haven't tested it yet

Tested and working! Note that the browser spinner will continue to spin while the data is refetched, but rendering the page is no longer blocked. Presumably, the browser spinner keeps spinning for as long as the stream is open, so I think this is expected.
